lib/services/local/headlessauthn_test.go

```diff
@@ -0,0 +1,156 @@
+		createStubErr     require.ErrorAssertionFunc
+		compareAndSwapErr require.ErrorAssertionFunc
+	}{
+		{
+			name: "OK headless authentication",
+			ha: &types.HeadlessAuthentication{
+				ResourceHeader: types.ResourceHeader{
+					Metadata: types.Metadata{
+						Name:    pubUUID,
+						Expires: &expires,
+					},
+				},
+				User:      "user",
+				PublicKey: []byte(sshPubKey),
+			},
+		}, {
+			name: "NOK name missing",
+			ha: &types.HeadlessAuthentication{
+				ResourceHeader: types.ResourceHeader{
+					Metadata: types.Metadata{
+						Expires: &expires,
+					},
+				},
+				User:      "user",
+				PublicKey: []byte(sshPubKey),
+			},
+			createStubErr: expectBadParameter,
+		}, {
+			name: "NOK expires missing",
+			ha: &types.HeadlessAuthentication{
+				ResourceHeader: types.ResourceHeader{
+					Metadata: types.Metadata{
+						Name: pubUUID,
+					},
+				},
+				User:      "user",
+				PublicKey: []byte(sshPubKey),
+			},
+			compareAndSwapErr: expectBadParameter,
+		}, {
+			name: "NOK username missing",
+			ha: &types.HeadlessAuthentication{
+				ResourceHeader: types.ResourceHeader{
+					Metadata: types.Metadata{
+						Name:    pubUUID,
+						Expires: &expires,
+					},
+				},
+				PublicKey: []byte(sshPubKey),
+			},
+			compareAndSwapErr: expectBadParameter,
+		}, {
+			name: "NOK name not derived from public key",
+			ha: &types.HeadlessAuthentication{
+				ResourceHeader: types.ResourceHeader{
+					Metadata: types.Metadata{
+						Name:    uuid.NewString(),
+						Expires: &expires,
+					},
+				},
+				User:      "user",
+				PublicKey: []byte(sshPubKey),
+			},
+			compareAndSwapErr: expectBadParameter,
+		},
+	}
+
+	for _, test := range tests {
+		t.Run(test.name, func(t *testing.T) {
+			stub, err := identity.CreateHeadlessAuthenticationStub(ctx, test.ha.Metadata.Name)
+			if test.createStubErr != nil {
+				test.createStubErr(t, err)
+				return
+			}
+			require.NoError(t, err, "CreateHeadlessAuthenticationStub returned non-nil error")
+
+			t.Cleanup(func() {
+				err = identity.DeleteHeadlessAuthentication(ctx, test.ha.Metadata.Name)
+				require.NoError(t, err)
+
+				_, err = identity.GetHeadlessAuthentication(ctx, test.ha.Metadata.Name)
+				require.True(t, trace.IsNotFound(err), "expected not found error but got: %v", err)
+			})
+
+			swapped, err := identity.CompareAndSwapHeadlessAuthentication(ctx, stub, test.ha)
+			if test.compareAndSwapErr != nil {
+				test.compareAndSwapErr(t, err)
+				return
+			}
+			require.NoError(t, err, "CompareAndSwapHeadlessAuthentication returned non-nil error")
+
+			retrieved, err := identity.GetHeadlessAuthentication(ctx, test.ha.Metadata.Name)
+			require.NoError(t, err, "GetHeadlessAuthentication returned non-nil error")
+			require.Equal(t, swapped, retrieved)
+		})
+	}
+}
+
+// sshPubKey is a randomly-generated public key used for login tests.
+//
+// The corresponding private key is:
+// -----BEGIN PRIVATE KEY-----
```

Comment:
People run scanners on OSS code all the time and we're likely to get a bunch of false reports of credential exposure here. Do we actually need the private key for anything? Can people just generate a new keypair if they suddenly have a need for the private key?